pygfx/objects/_events.py

```diff
@@ -0,0 +1,419 @@
+from collections import defaultdict
+from enum import Enum
+import logging
+from time import perf_counter_ns
+from typing import Union
+from weakref import ref, WeakValueDictionary
+
+
+logger = logging.getLogger("pygfx")
+
+err_hashes = {}
+
+
+def log_exception(kind, err):
+    """Log the given exception instance, but only log a one-liner for
+    subsequent occurances of the same error to avoid spamming (which
+    can happen easily with errors in the event handlers).
+    """
+    msg = str(err)
+    msgh = hash(msg)
+    if msgh not in err_hashes:
+        # Provide the exception, so the default logger prints a stacktrace.
+        # IDE's can get the exception from the root logger for PM debugging.
+        err_hashes[msgh] = 1
+        logger.error(kind, exc_info=err)
+    else:
+        # We've seen this message before, return a one-liner instead.
+        err_hashes[msgh] = count = err_hashes[msgh] + 1
+        msg = kind + ": " + msg.split("\n")[0].strip()
+        msg = msg if len(msg) <= 70 else msg[:69] + "…"
+        logger.error(msg + f" ({count})")
+
+
+CLICK_DEBOUNCE = 500  # in milliseconds
+
+
+class EventType(str, Enum):
+    # Keyboard
+    KEY_DOWN = "key_down"
+    KEY_UP = "key_up"
+    # Pointer
+    POINTER_DOWN = "pointer_down"
+    POINTER_MOVE = "pointer_move"
+    POINTER_UP = "pointer_up"
+    CLICK = "click"
+    DOUBLE_CLICK = "double_click"
+    # Wheel
+    WHEEL = "wheel"
+    # Window
+    CLOSE = "close"
+    RESIZE = "resize"
+
+
+class Event:
+    """Event base class for creating events.
+
+    If a target is set, an event can bubble up through a hierarchy
+    of targets, connected through a ``parent`` property.
+    To prevent an event from bubbling up, use ``stop_propagation``.
+
+    It is also possible to cancel events, which will stop any further
+    handling of the event (also by the same target).
+    """
+
+    def __init__(
+        self,
+        type: Union[str, EventType],
+        *,
+        bubbles=True,
+        target: "EventTarget" = None,
+        time_stamp: float = None,
+        cancelled: bool = False,
+        **kwargs,
```

Comment:
Kwargs come in now, but are simply ignored.
```suggestion
```

Reply:
I do use this functionality to convert from `wgpu-py` events to these events, so I would like to keep this in here.

Reply:
Fixed by adding just the `event_type` argument that will be ignored instead of every kwarg.